contracts/minter-suite/Minters/MinterSEAV0.sol

```diff
@@ -0,0 +1,987 @@
+     * `msg.sender`. A fallback method of sending funds back to the bidder via
+     * WETH is used if the bidder address is not accepting ETH (preventing
+     * denial of service attacks).
+     * Note that the use of `_tokenId` is to prevent the possibility of
+     * transactions that are stuck in the pending pool for long periods of time
+     * from unintentionally bidding on auctions for future tokens.
+     * @param _tokenId Token ID being bidded on
+     */
+    function createBid(uint256 _tokenId) external payable {
+        createBid_4cM(_tokenId);
+    }
+
+    /**
+     * @notice gas-optimized version of createBid(uint256,uint256).
+     * @dev nonReentrant modifier is used to prevent reentrancy attacks, e.g.
+     * an an auto-bidder that would be able to atomically outbid a user's
+     * new bid via a reentrant call to createBid.
+     */
+    function createBid_4cM(uint256 _tokenId) public payable nonReentrant {
+        uint256 _projectId = _tokenId / ONE_MILLION;
+        // CHECKS
+        ProjectConfig storage _projectConfig = projectConfig[_projectId];
+        Auction storage _auction = _projectConfig.activeAuction;
+
+        // if no auction exists, or current auction is already settled, attempt
+        // to initialize a new auction for the input token ID and immediately
+        // return
+        if ((!_auction.initialized) || _auction.settled) {
+            _initializeAuction(_projectId, _tokenId);
+            return;
+        }
+        // @dev this branch is guaranteed to have an initialized auction that
+        // not settled, so no need to check for initialized or not settled
+
+        // ensure bids for a specific token ID are only applied to the auction
+        // for that token ID.
+        require(
+            _auction.tokenId == _tokenId,
+            "Token ID does not match auction"
+        );
+
+        // ensure auction is not already ended
+        require(_auction.endTime > block.timestamp, "Auction already ended");
+
+        // require bid to be sufficiently greater than current highest bid
+        // @dev no overflow enforced automatically by solidity ^8.0.0
+        require(
+            msg.value >=
+                (_auction.currentBid *
+                    (100 + minterMinBidIncrementPercentage)) /
+                    100,
+            "Bid is too low"
+        );
+
+        // EFFECTS
+        // record previous highest bid details for refunding
+        uint256 previousBid = _auction.currentBid;
+        address payable previousBidder = _auction.currentBidder;
+
+        // update auction state
+        _auction.currentBid = msg.value;
+        _auction.currentBidder = payable(msg.sender);
+        uint256 minEndTime = block.timestamp + minterTimeBufferSeconds;
+        if (_auction.endTime < minEndTime) {
+            _auction.endTime = minEndTime.toUint64();
+        }
+
+        // INTERACTIONS
+        // refund previous highest bidder
```

Comment:
more of a design question and definitely not blocking: were user claim/settlement style bid refunds considered? Aside from reducing gas cost of this bidding, this would also potentially allow for an implementation where someone that is outbid can add create a new active bid by adding funds to a previous bid that was outbid but has not been claimed and is still being held by the contract. Totally understand if the thinking here is that the risk of holding funds outweighs the potential mentioned benefits though.

Reply:
This is a very interesting design consideration. I do think though that the primary value proposition in this case though would just be gas costs, no?

> this would also potentially allow for an implementation where someone that is outbid can add create a new active bid by adding funds to a previous bid that was outbid but has not been claimed and is still being held by the contract. 

Put differently, what is the benefit of this itself other than the gas savings angle? The rebidding flow is still there as an option either way, and needs to have explicit user action associated with it.

Reply:
agree,  practically speaking it all boils down to gas cost savings. The rebidding flow is there regardless and explicit regardess, but with this design the rebidding flow maybe has a different psychological user experience (adding a small amount to get back "in the game" vs re-adding the entire amount from your previous bid + amount to make your new bid valid)

Reply:
Good discussion topic here.

Re: gas efficiency, I think the current implementation is actually slightly more globally gas-efficient, because it will require at least one less tx per auction (the second-place bidder doesn't have to reclaim their ETH at the end of an auction), and each tx has a base cost of 21000 gas.

I think probably more important than gas efficiency is UX. While it can be debated that a pull mechanism is sometimes preferred relative to a push, I think following the nouns team's implementation of a push mechanism is slightly more user friendly for this type of auction. The single biggest benefit of immediately refunding the losing bidder is that it eliminates the need for a user to worry about the question "Do I have any funds to go reclaim in Art Blocks auctions?". It also eliminates the need for us to design and implement an entire notification system for a user to communicate which project(s) they have funds available to reclaim on.

These are just my instincts though, and definitely consider this topic worth discussing further across teams!

Reply:
Ah ya good points on the global gas efficiency + extra work around notification system 👍 

Though global gas efficiency aside, I would say it's potentially nicer if the refund gas costs are associated/billed to the bidder receiving the refund vs a different bidder entirely. That being said, the points mentioned above are pretty compelling to the side of it being not worth it

Reply:
>also just feels not worth it for locking funds

This is a very good point 😅

Reply:
I know we had talked about this more synchronously since this async convo, but upon more conversation definitely understand how the variation suggested here could have a difference in perceptual effect when interacting with this minter. That said, for our initial MVP of this minter I don't think the additional complexity that this would introduce is warranted given that the potential perceptual benefits are definitely hypothetical and could always be introduced via an iterative variant if we had signal that there was demand to warrant this complexity in the future.